internal/controller/apiextensions/claim/configurator.go

```diff
@@ -107,3 +118,87 @@ func filter(in map[string]interface{}, keys ...string) map[string]interface{} {
 	}
 	return out
 }
+
+// APIClaimConfigurator configures the supplied claims with fields
+// from the composite. This includes late-initializing spec values
+// and updating status fields in claim.
+type APIClaimConfigurator struct {
+	client client.Client
+}
+
+// NewAPIClaimConfigurator returns a APIClaimConfigurator.
+func NewAPIClaimConfigurator(client client.Client) *APIClaimConfigurator {
+	return &APIClaimConfigurator{client: client}
+}
+
+// Configure the supplied claims with fields from the composite.
+// This includes late-initializing spec values and updating status fields in claim.
+func (c *APIClaimConfigurator) Configure(ctx context.Context, cr resource.CompositeClaim, cp resource.Composite) error {
+	ucr, ok := cr.(*claim.Unstructured)
+	if !ok {
+		return nil
+	}
+	ucp, ok := cp.(*composite.Unstructured)
+	if !ok {
+		return nil
+	}
+
+	if err := merge(ucr.Object["status"], ucp.Object["status"],
+		// Status fields from composite overwrite non-empty fields in claim
+		withMergeOptions(mergo.WithOverride),
+		withSrcFilter(xcrd.GetPropFields(xcrd.CompositeResourceStatusProps())...)); err != nil {
+		return errors.Wrap(err, errMergeClaimStatus)
+	}
+
+	if err := c.client.Status().Update(ctx, cr); err != nil {
+		return errors.Wrap(err, errUpdateClaimStatus)
+	}
+
+	if err := merge(ucr.Object["spec"], ucp.Object["spec"],
+		withSrcFilter(xcrd.GetPropFields(xcrd.CompositeResourceSpecProps())...)); err != nil {
+		return errors.Wrap(err, errMergeClaimSpec)
+	}
+
+	return errors.Wrap(c.client.Update(ctx, cr), errUpdateClaim)
+}
+
+// Note(mcavoyk): merge package might be a better fit in crossplane-runtime
```

Comment:
Historically I'd agree with you, but we have a few libs in runtime now that are _only_ used by the Composition controllers and it's kind of a pain to have to update runtime just to get something done in core. I'm starting to come around to the idea that we shouldn't put things in runtime until they have more than one consumer (e.g. they're used by multiple providers).